doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+- RDM: rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md
+- Title: Runtime configuration architecture
+- Authors: José Álamos, Leandro Lanzieri
+- Status: draft
+- Type: Design
+- Created: December 2018
+
+## Abstract
+This memo describes the proposed high-level architecture and
+mechanisms to implement a runtime configuration system on a RIOT node.
+
+A runtime configuration system is in charge of providing a mechanism to set and
+get the values of configuration parameters that are used during the execution
+of the firmware, as well as a way to persist these values. Runtime
+configurations are deployment-specific and can be changed on a per node basis.
+Appropriate management tools could also enable the configuration of node
+groups.
+
+Examples of runtime configurations are:
+- Transmission duty cycles
+- Sensor thresholds
+- Security credentials
+- System state variables
+
+These parameters might have constraints, like a specific order to be applied
+(due to interdependencies) or value boundaries.
```

Comment:
Can you provide a concrete example where an application programmer exposes configuration parameters from other modules in this context?

E.g at RH `set` level, most of the times would look like:
```
/* Application RH */
set(key, value) {
    if(key == "foo")
        internal_static_variable = value;
}
```

If the module wants to change e.g AT86RF settings, it would need to add functions to expose static variables, which is non-sense IMO. If an application wants to expose these configs, it should expose "Application" and "AT86RF" Registry Handlers.

Do you think it's needed to state this?

Reply:
@danpetry are you heading for a multi-stakeholder system of rights? I'm not sure we want to impose this complexity?

Reply:
> Are non-static configuration parameters possible/likely?

It's always possible, but I think defining how the configurations should be implemented is out of the scope of this document. We don't have any mechanism in RIOT to protect modules (we have no users, groups, etc). So, it will be always possible to mess with other components (not only in the RCS scope)

Reply:
ok, yes I see what you're saying, agreed